search/engine.go

```diff
@@ -559,17 +561,77 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 		}
 	}
 
+	var currentLang string
 	results := make([]*elastic.SearchResult, 0)
 	for _, lang := range query.LanguageOrder {
 		if r, ok := resultsByLang[lang]; ok {
 			results = r
+			currentLang = lang
 			break
 		}
 	}
 
 	ret, err := joinResponses(sortBy, from, size, results...)
 
-	if ret != nil && ret.Hits != nil {
+	if ret != nil && ret.Hits != nil && ret.Hits.Hits != nil {
+
+		log.Info("Preparing highlights search.")
+		mssHighlights := e.esc.MultiSearch()
+
+		for _, h := range ret.Hits.Hits {
+
+			if h.Id == "" || strings.HasPrefix(h.Index, "intent-") {
+				//  Bypass intent
+				continue
+			}
+
+			mssHighlights.Add(NewResultsSearchRequest(
+				SearchRequestOptions{
+					resultTypes:      consts.ES_SEARCH_RESULT_TYPES,
+					docIds:           []string{h.Id},
+					index:            h.Index,
+					query:            Query{ExactTerms: query.ExactTerms, Term: query.Term, Filters: query.Filters, LanguageOrder: []string{currentLang}},
+					sortBy:           consts.SORT_BY_RELEVANCE,
+					from:             0,
+					size:             1,
+					preference:       preference,
+					useHighlight:     true,
+					partialHighlight: true}))
+		}
+
+		log.Info("Searching for highlights and replacing original results with highlighted results.")
+
+		beforeHighlightsDoSearch := time.Now()
+		mr, err := mssHighlights.Do(context.TODO())
+		e.timeTrack(beforeHighlightsDoSearch, "DoSearch.MultisearcHighlightsDo")
+		if err != nil {
+			return nil, errors.Wrap(err, "ESEngine.DoSearch - Error mssHighlights Do.")
+		}
+
+		for _, highlightedResults := range mr.Responses {
+			if highlightedResults.Error != nil {
+				log.Warnf("%+v", highlightedResults.Error)
+				return nil, errors.New(fmt.Sprintf("Failed multi get highlights: %+v", highlightedResults.Error))
+			}
+			if haveHits(highlightedResults) {
```

Comment:
It is important to validate this search always succeeds. In case of failure we do want to log this in production as we expect 100% of searches to succeed.